packages/jest-resolve/src/index.js

```diff
@@ -134,23 +136,29 @@ class Resolver {
     }
 
     // 0. If we have already resolved this module for this directory name,
-    //    return a value from the cache.
+    // return a value from the cache.
     if (this._moduleNameCache[key]) {
       return this._moduleNameCache[key];
     }
 
     // 1. Check if the module is a haste module.
-    let module = this.getModule(moduleName);
+    module = this.getModule(moduleName);
+    if (module) {
+      return (this._moduleNameCache[key] = module);
+    }
+
+    // 2. Check if the module is a haste module (but now, in the mocks).
+    module = this.getMockModule(dirname + path.sep + 'index.js', moduleName);
```

Comment:
Have you considered how (if at all) this should interact with the `paths` option to `resolve`?

Reply:
Yep; this replicates the step 1, but using the mock registry of the Haste map, instead of the module registry. However, it goes after 1, since, between a file and a mock, the file should always be returned by `require.resolve`, even if mocked (deciding the responsibility of what object to return is part of `require`, not `require.resolve`).

Adding this extra step aims to be able to resolve Haste mocks which do not have a real file counter part; e.g.:

```
+ Foo.js
+ Bar.js
+ __mocks__
   + Foo.js
   + Baz.js
```

In the above example, `require('Foo')`, `require('Bar')` and `require('Baz')` all work (when using Haste); however only `require.resolve('Foo')` and `require.resolve('Bar')` work. Also,  `require.resolve('Foo')` returns the root path, even if `Foo` is mocked.

Reply:
Sorry if I was unclear, I meant something like `require.resolve('./mod', {paths: ['../dir']})`

Reply:
`paths` are not used for Haste resolutions :)